win32ss/gdi/dib/dib32bpp.c

```diff
@@ -99,92 +196,211 @@ DIB_32BPP_BitBltSrcCopy(PBLTINFO BltInfo)
-        SourceBits += 2;
+        if (bLeftToRight)
+        {
+          SourceBits -= 2;
+        }
+        else
+        {
+          SourceBits += 2;
+        }
         DestBits += 4;
       }
 
-      SourceLine += BltInfo->SourceSurface->lDelta;
+      if (bTopToBottom)
+      {
+        SourceLine -= BltInfo->SourceSurface->lDelta;
+      }
+      else
+      {
+        SourceLine += BltInfo->SourceSurface->lDelta;
+      }
       DestLine += BltInfo->DestSurface->lDelta;
     }
     break;
 
   case BMF_24BPP:
+    DPRINT("24BPP Case Selected with DestRect Width of '%d'.\n",
+      BltInfo->DestRect.right - BltInfo->DestRect.left);
+
+    /* This sets SourceLine to the top line */
     SourceLine = (PBYTE)BltInfo->SourceSurface->pvScan0
       + (BltInfo->SourcePoint.y * BltInfo->SourceSurface->lDelta)
       + 3 * BltInfo->SourcePoint.x;
+
+    if (bTopToBottom)
+    {
+      /* This sets SourceLine to the bottom line */
+      SourceLine += BltInfo->SourceSurface->lDelta * (BltInfo->DestRect.bottom - BltInfo->DestRect.top - 1);
+    }
+
     DestLine = DestBits;
 
     for (j = BltInfo->DestRect.top; j < BltInfo->DestRect.bottom; j++)
     {
       SourceBits = SourceLine;
       DestBits = DestLine;
 
+      if (bLeftToRight)
+      {
+        /* This sets the SourceBits to the rightmost pixel */
+        SourceBits += (BltInfo->DestRect.right - BltInfo->DestRect.left - 1) * 3;
+      }
+
       for (i = BltInfo->DestRect.left; i < BltInfo->DestRect.right; i++)
       {
         xColor = (*(SourceBits + 2) << 0x10) +
           (*(SourceBits + 1) << 0x08) +
           (*(SourceBits));
         *((PDWORD)DestBits) = (DWORD)XLATEOBJ_iXlate(BltInfo->XlateSourceToDest, xColor);
-        SourceBits += 3;
+        if (bLeftToRight)
+        {
+          SourceBits -= 3;
+        }
+        else
+        {
+          SourceBits += 3;
+        }
         DestBits += 4;
       }
 
-      SourceLine += BltInfo->SourceSurface->lDelta;
+      if (bTopToBottom)
+      {
+        SourceLine -= BltInfo->SourceSurface->lDelta;
+      }
+      else
+      {
+        SourceLine += BltInfo->SourceSurface->lDelta;
+      }
       DestLine += BltInfo->DestSurface->lDelta;
     }
     break;
 
   case BMF_32BPP:
-    if (NULL == BltInfo->XlateSourceToDest ||
-      0 != (BltInfo->XlateSourceToDest->flXlate & XO_TRIVIAL))
+    DPRINT("32BPP Case Selected with SrcPt (%d,%d) and DestRect Width/height of '%d/%d'.\n",
+      BltInfo->SourcePoint.x, BltInfo->SourcePoint.y,
+      BltInfo->DestRect.right - BltInfo->DestRect.left,
+      BltInfo->DestRect.bottom - BltInfo->DestRect.top);
+
+    /* This tests for whether we can use simplified/quicker code below which uses RtlMoveMemory.
+     * It works for increasing source and destination areas only where there is no full overlap and no flip.
+     */
+    if ((NULL == BltInfo->XlateSourceToDest || 0 != (BltInfo->XlateSourceToDest->flXlate & XO_TRIVIAL))
+      && (!bTopToBottom && !bLeftToRight))
```

Comment:
Since you are changing this line, please fix inverse logic.
```suggestion
    if ((BltInfo->XlateSourceToDest == NULL ||
      (BltInfo->XlateSourceToDest->flXlate & XO_TRIVIAL) != 0) &&
      (!bTopToBottom && !bLeftToRight))
```